cirq/contrib/quimb/mps_simulator.py

```diff
@@ -231,92 +235,243 @@ def sample(
         return self.state_vector()
 
     def apply_unitary(self, op: 'cirq.Operation'):
-        idx = [self.qubit_map[qubit] for qubit in op.qubits]
         U = protocols.unitary(op).reshape([qubit.dimension for qubit in op.qubits] * 2)
 
-        if len(idx) == 1:
-            n = idx[0]
-            self.M[n] = np.einsum('ij,mnj->mni', U, self.M[n])
-        elif len(idx) == 2:
-            if abs(idx[0] - idx[1]) != 1:
-                raise ValueError('Can only handle continguous qubits')
-            elif idx[0] < idx[1]:
-                n, p = idx
-            else:
-                p, n = idx
-                U = np.swapaxes(np.swapaxes(U, 0, 1), 2, 3)
-            T = np.einsum('klij,mni,npj->mkpl', U, self.M[n], self.M[p])
-            X, S, Y = np.linalg.svd(T.reshape([T.shape[0] * T.shape[1], T.shape[2] * T.shape[3]]))
-            X = X.reshape([T.shape[0], T.shape[1], -1])
-            Y = Y.reshape([-1, T.shape[2], T.shape[3]])
-
-            S = np.asarray([math.sqrt(x) for x in S])
-
-            nkeep = 0
-            for i in range(S.shape[0]):
-                if S[i] >= S[0] * self.threshold:
-                    nkeep = i + 1
-
-            X = X[:, :, :nkeep]
-            S = np.diag(S[:nkeep])
-            Y = Y[:nkeep, :, :]
-
-            self.M[n] = np.einsum('mis,sn->mni', X, S)
-            self.M[p] = np.einsum('ns,spj->npj', S, Y)
+        if len(op.qubits) == 1:
+            n = self.qubit_map[op.qubits[0]]
+
+            old_n = self.i_str(n)
+            new_n = 'new_' + old_n
+
+            U = qtn.Tensor(U, inds=(new_n, old_n))
+            self.M[n] = (U @ self.M[n]).reindex({new_n: old_n})
+        elif len(op.qubits) == 2:
+            self.num_svd_splits += 1
+
+            n, p = [self.qubit_map[qubit] for qubit in op.qubits]
+
+            old_n = self.i_str(n)
+            old_p = self.i_str(p)
+            new_n = 'new_' + old_n
+            new_p = 'new_' + old_p
+
+            U = qtn.Tensor(U, inds=(new_n, new_p, old_n, old_p))
+
+            # This is the index on which we do the joining. We need to add it iff it's the first
+            # time that we do the joining for that specific pair.
+            mu_ind = self.mu_str(n, p)
+            if mu_ind not in self.M[n].inds:
+                self.M[n].new_ind(mu_ind)
+            if mu_ind not in self.M[p].inds:
+                self.M[p].new_ind(mu_ind)
+
+            T = U @ self.M[n] @ self.M[p]
+
+            left_inds = tuple(set(T.inds) & set(self.M[n].inds)) + (new_n,)
+            X, Y = T.split(
+                left_inds,
+                cutoff=self.rsum2_cutoff,
+                cutoff_mode='rsum2',
+                get='tensors',
+                absorb='both',
+                bond_ind=mu_ind,
+            )
+
+            self.M[n] = X.reindex({new_n: old_n})
+            self.M[p] = Y.reindex({new_p: old_p})
         else:
+            # NOTE(tonybruguier): There could be a way to handle higher orders. I think this could
```

Comment:
I think `T.split` will work for higher-order tensors.

Reply:
I am a bit confused by your comment, but I did take a second look at the code:
https://sourcegraph.com/github.com/jcmgray/quimb@88cc3d88c2a43f08e95b32528594e3c2239d2184/-/blob/quimb/tensor/tensor_core.py#L611

Let's say that we have a 3-qubit gate, shouldn't we split along three groups of indices? Like instead of 'left_inds' and 'right_inds', we would have 'group_1_inds', 'group_2_inds', and 'group_3_inds'.

Leaving unresolved, because I suspect I am missing something.

Reply:
This is outside of the scope of the current PR, and I'm not sure I understand 100% the algorithm; but you can still split it either asymmetrically (and just accept that entanglement will grow, albeit slower than if you kept it as 3 qubits). or split twice (i.e. SVD into (1qubit)(singularvalues)(2qubit) --> (1qubit)(singularvalues1)(1qubit)(singularvalues2)(1qubit))

Reply:
Ack.